File: frontend/src/coparent/components/calendar/recurrence.ts
import type { Event } from '../../types/calendar';

const WEEKDAY_KEYS = ['sunday', 'monday', 'tuesday', 'wednesday', 'thursday', 'friday', 'saturday'];

const toYmd = (value: string) => (value.includes('T') ? value.slice(0, 10) : value);
const toLocalDay = (value: string) => new Date(`${toYmd(value)}T12:00:00`);

const addDays = (date: Date, amount: number) => {
  const next = new Date(date);
  next.setDate(next.getDate() + amount);
  return next;
};

export const dateToYmd = (date: Date) => {
  const year = date.getFullYear();
  const month = String(date.getMonth() + 1).padStart(2, '0');
  const day = String(date.getDate()).padStart(2, '0');
  return `${year}-${month}-${day}`;
};

// Two letters identify every weekday, so any spelling from "TU" to "Tuesday" resolves to the
// one key the expansion compares against. The backend stores canonical names; this keeps a
// restored or older series from being saved yet silently rendering no occurrences.
const weekdayKey = (value: string) => {
  const lower = value.trim().toLowerCase();
  return lower.length >= 2 ? WEEKDAY_KEYS.find((key) => key.startsWith(lower)) : undefined;
};

export const expandRecurringEvents = (events: Event[], rangeStart: string, rangeEnd: string) => {
  const rangeStartDay = toLocalDay(rangeStart);
  const rangeEndDay = toLocalDay(rangeEnd);

  return events.flatMap((event) => {
    if (!event.recurring) return [event];

    const eventStart = toLocalDay(event.startDate);
    // A series with no end date repeats indefinitely; treating the start as its end would
    // show one occurrence and nothing after it.
    const eventEnd = event.endDate ? toLocalDay(event.endDate) : rangeEndDay;
    const effectiveStart = eventStart > rangeStartDay ? eventStart : rangeStartDay;
    const effectiveEnd = eventEnd < rangeEndDay ? eventEnd : rangeEndDay;

    if (effectiveStart > effectiveEnd) return [];

    const durationDays = event.endDate
      ? Math.max(
          0,
          Math.round((eventEnd.getTime() - eventStart.getTime()) / (1000 * 60 * 60 * 24)),
        )
      : 0;

    const frequency = event.recurring.frequency;
    const skipped = new Set(event.recurring.excludedDates ?? []);

    if (frequency === 'daily') {
      const days: Event[] = [];
      for (let day = new Date(effectiveStart); day <= effectiveEnd; day = addDays(day, 1)) {
        const startYmd = dateToYmd(day);
        if (skipped.has(startYmd)) continue;
        const endYmd = dateToYmd(addDays(day, durationDays));
        days.push({
          ...event,
          id: `${event.id}:${startYmd}`,
          sourceId: event.id,
          startDate: startYmd,
          endDate: endYmd,
        });
      }
      return days;
    }

    if (frequency === 'weekly') {
      const recognisedDays = (event.recurring.days ?? []).flatMap((day) => {
        const key = weekdayKey(day);
        return key ? [key] : [];
      });
      const allowedDays =
        recognisedDays.length > 0 ? recognisedDays : [WEEKDAY_KEYS[eventStart.getDay()]];

      const days: Event[] = [];
      for (let day = new Date(effectiveStart); day <= effectiveEnd; day = addDays(day, 1)) {
        const weekday = WEEKDAY_KEYS[day.getDay()];
        if (!allowedDays.includes(weekday)) continue;

        const startYmd = dateToYmd(day);
        if (skipped.has(startYmd)) continue;
        const endYmd = dateToYmd(addDays(day, durationDays));
        days.push({
          ...event,
          id: `${event.id}:${startYmd}`,
          sourceId: event.id,
          startDate: startYmd,
          endDate: endYmd,
        });
      }
      return days;
    }

    return [event];
  });
};

/**
 * What a click on a rendered event refers to: the stored event, plus the date of the
 * occurrence when it came from a repeating series, so that one week can be skipped.
 */
export const occurrenceTarget = (event: Event): [string, string | undefined] =>
  event.sourceId ? [event.sourceId, event.startDate] : [event.id, undefined];

File: backend/src/main/java/com/simonrowe/coparent/calendar/Recurrence.java
package com.simonrowe.coparent.calendar;

import java.util.List;
import java.util.Locale;
import java.util.Optional;

/**
 * The one recurrence vocabulary the calendar can render.
 *
 * <p>The browser expands a weekly event by comparing each day against these exact lowercase
 * names, so a series stored as {@code "TUE"} or {@code "Tuesday"} is saved successfully and then
 * produces no occurrences at all: it exists and is never shown. Every writer therefore stores
 * the canonical names, and the domain refuses anything else.
 */
public final class Recurrence {

  public static final List<String> FREQUENCIES = List.of("daily", "weekly");

  public static final List<String> DAYS = List.of(
      "monday", "tuesday", "wednesday", "thursday", "friday", "saturday", "sunday");

  private Recurrence() {
  }

  /**
   * Maps a weekday spelling to its canonical name. Any prefix of at least two letters is
   * accepted ({@code TU}, {@code tue}, {@code Tues}, {@code Tuesday}); two letters already
   * identify every weekday uniquely, so no prefix is ambiguous.
   */
  public static Optional<String> canonicalDay(final String value) {
    if (value == null) {
      return Optional.empty();
    }
    final String lower = value.trim().toLowerCase(Locale.ROOT);
    if (lower.length() < 2) {
      return Optional.empty();
    }
    return DAYS.stream().filter(day -> day.startsWith(lower)).findFirst();
  }

  /** Maps a frequency spelling to its canonical name, or empty when it is not supported. */
  public static Optional<String> canonicalFrequency(final String value) {
    if (value == null) {
      return Optional.empty();
    }
    final String lower = value.trim().toLowerCase(Locale.ROOT);
    return FREQUENCIES.contains(lower) ? Optional.of(lower) : Optional.empty();
  }
}
